### PR TITLE
Implement the Zora monad

### PR DESCRIPTION
**Pull request by PureFunctor**
This implements the Zora monad, which is an abstraction that allows for `Effect` to be ran with `ST`, and elided if needed.

### CODE REVIEW

**Review by mikesol — APPROVED** [review]
This looks really nice! The only question I'd have is if it's possible to keep a generic `fromEvent` and `toEvent` (in addition to the ones you've provided which are great) or if those need to be dropped now.

Let's let @MonoidMusician look at it when she's back from 🌴🍹 and if it gets the nod then 🚀

**Comment by mikesol** [comment]
If it's note done already, @PureFunctor could you also run the test suite against `AnEvent Zora`?

**Comment by PureFunctor** [comment]
I could also port over the tests I have for `Mermaid`

File: test/Main.purs [hunk 1]
```diff
@@ -215,6 +216,13 @@ main = do
                   Ref.read rf >>= shouldEqual x
                   unsub
         suite "Event" (\i f -> f i) Event.create Event.subscribe
+        let
+          zoraCreate :: forall a. Effect { event :: AnEvent Zora a , push :: a -> Effect Unit }
+          zoraCreate = Event.create <#> \r -> r { event = fromEvent r.event }
+
+          zoraSubscribe :: forall a. AnEvent Zora a -> (a -> Effect Unit) -> Effect (Effect Unit)
+          zoraSubscribe = Event.subscribe <<< toEvent
```
Comment (PureFunctor) [inline test/Main.purs:224]: While these don't test `Zora` as it is, it mirrors more of its real-world use-cases where `fromEvent`/`toEvent` has to be invoked, like with `deku`'s `bus`.

**Comment by PureFunctor** [comment]
> This looks really nice! The only question I'd have is if it's possible to keep a generic `fromEvent` and `toEvent` (in addition to the ones you've provided which are great) or if those need to be dropped now.

Is it used in any other context aside from SSR through Deku? I'm keen on dropping them if this is the case.

**Comment by mikesol** [comment]
Nice! Once it gets a second review we can merge.

**Review by MonoidMusician — APPROVED** [review]
This looks fine to me. I think `fromStEvent` is dodgy and maybe we should remove it for now – I don’t think it will end up being used in Deku or apps built with Deku, right? They should either be building `Effect` or `Zora` events specifically or working with polymorphic `MonadST`, they shouldn’t build a monomorphic `StEvent` and hope to convert it later …

I think if we really wanted more freedom to make this work we would have to separate it into a profunctor setup with an input monad and an output monad, `ProEvent m n a = ((a -> m Unit) -> n (n Unit)`, and that might allow you to preserve callbacks. But no sense being more polymorphic when less will do!

**Comment by mikesol** [comment]
Great! Many thanks to @PureFunctor for this amazing work. Now that it's approved, my plan is to make a minor-version-bump of `hyrule`, `bolson`, `deku`, `ocarina` and `rito` tonight and also to make update PRs to `paraglider`, `joyride`, `package-sets` and any other relevant libs.

**Comment by PureFunctor** [comment]
I'll fix the conflict in a few so we can get this merged soon. I'm keen on keeping the ST stuff in for posterity.

**Comment by mikesol** [comment]
> I'll fix the conflict in a few so we can get this merged soon. I'm keen on keeping the ST stuff in for posterity.

Haha, beat you to it :-D

**Comment by mikesol** [comment]
🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Pull request merged**